{"category": "ham", "to_address": "r-help@stat.math.ethz.ch", "from_address": "Skip Montanaro <skip@pobox.com>", "subject": "[R] relocation error in grDevices.so", "body": "(Warning: I'm not an R guy.  I'm a Python guy trying to get the\nR-Python interface working again after some upgrades.)\n\nI'm trying to upgrade our numpy/rpy/matplotlib environment (Solaris\n10/Intel, Python 2.4).  In the process I found I needed to rebuild R\n(2.1.1) because it was compiled with gcc 3.3.2 and we have since\nmigrated to gcc 3.4.1.  I'm using this configure setup:\n\n    export CPPFLAGS=-I/opt/app/include\n    export LDFLAGS='-L/opt/app/lib -R/opt/app/lib'\n    ./configure --prefix=/opt/app/R-2.1.1-gcc-3.4.1\n\nAfter building and installing R I downloaded and installed the\ncontributed quadprog library.  I ran this simple solver example using\nPython+rpy which works (well, except for some R error I haven't\nfigured out yet) in the old installation:\n\n    import numpy\n    import rpy\n    rpy.r(\"library(quadprog)\")\n\n    Dmat = numpy.identity(3, numpy.float_)\n    print Dmat\n    rpy.r.assign(\"Dmat\", Dmat)\n\n    dvec = numpy.array([0,5,0], numpy.float_)\n    print dvec\n    rpy.r.assign(\"dvec\", dvec)\n\n    Amat = numpy.array([[-4,-3,0], [2,1,0], [0,-2,1]], numpy.float_)\n    print Amat\n    rpy.r.assign(\"Amat\", Amat.transpose())\n\n    bvec = numpy.array([-8,2,0], numpy.float_)\n    print bvec\n    rpy.r.assign(\"bvec\", bvec)\n\n    result = rpy.r(\"solve.QP(Dmat, dvec, Amat, bvec=bvec)\")\n    print result['solution']\n\nand get this error:\n\n    Error in dyn.load(x, as.logical(local), as.logical(now)) :\n\t    unable to load shared library\n'/opt/app/R-2.1.1-gcc-3.4.1/lib/R/library/grDevices/libs/grDevices.so':\n      ld.so.1: python: fatal: relocation error: file\n/opt/app/R-2.1.1-gcc-3.4.1/lib/R/library/grDevices/libs/grDevices.so: symbol\nR_NilValue: referenced symbol not found\n    Loading required package: grDevices\n    Error in dyn.load(x, as.logical(local), as.logical(now)) :\n\t    unable to load shared library\n'/opt/app/R-2.1.1-gcc-3.4.1/lib/R/library/grDevices/libs/grDevices.so':\n      ld.so.1: python: fatal: relocation error: file\n/opt/app/R-2.1.1-gcc-3.4.1/lib/R/library/grDevices/libs/grDevices.so: symbol\nR_NilValue: referenced symbol not found\n    In addition: Warning message:\n    package grDevices in options(\"defaultPackages\") was not found\n    Error: package 'grDevices' could not be loaded\n    Traceback (most recent call last):\n      File \"rpytest.py\", line 3, in ?\n\trpy.r(\"library(quadprog)\")\n      File\n\"/opt/app/g++lib6/python-2.4-pre-release/lib/python2.4/site-packages/rpy.py\",\nline 299, in __call__\n\treturn self.eval(self.parse(text=s))\n    rpy.RException: Error: package/namespace load failed for 'quadprog'\n\nLooking at grDevices.so I see that it wasn't even linked against the R\nlibrary:\n\n    % ldd /opt/app/R-2.1.1-gcc-3.4.1/lib/R/library/grDevices/libs/grDevices.so\n\t    libc.so.1 =>     /lib/libc.so.1\n\t    libm.so.2 =>     /lib/libm.so.2\n\nso I can't see how it could possibly resolve the R_NilValue symbol.\n\nCan someone tell me where I've gone awry?\n\nThanks,\n\nSkip Montanaro\nskip@pobox.com\n\n______________________________________________\nR-help@stat.math.ethz.ch mailing list\nhttps://stat.ethz.ch/mailman/listinfo/r-help\nPLEASE do read the posting guide http://www.R-project.org/posting-guide.html\nand provide commented, minimal, self-contained, reproducible code.\n\n"}